src/trace/log.zig

```diff
@@ -1,366 +1,491 @@
+};
 
-pub const Logger = union(enum) {
-    standard: *StandardErrLogger,
-    test_logger: TestLogger,
+const LogKind = enum {
+    standard,
+    test_logger,
     noop,
+};
 
-    const Self = @This();
-
-    pub fn init(allocator: std.mem.Allocator, max_level: Level) Self {
-        return .{ .standard = StandardErrLogger.init(allocator, max_level) };
-    }
-
-    pub fn spawn(self: Self) void {
-        switch (self) {
-            .standard => |logger| {
-                logger.spawn();
-            },
-            .noop, .test_logger => {},
-        }
-    }
-
-    pub fn deinit(self: Self) void {
-        switch (self) {
-            .standard => |logger| {
-                logger.deinit();
-            },
-            .noop, .test_logger => {},
-        }
-    }
-
-    pub fn field(self: Self, name: []const u8, value: anytype) Entry {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                return logger.field(name, value);
-            },
-            .noop => {
-                return .noop;
-            },
+/// A ScopedLogger could either be:
+/// - A StandardErrLogger
+/// - A TestingLogger
+pub fn ScopedLogger(comptime scope: ?[]const u8) type {
+    return union(LogKind) {
+        standard: *ChannelPrintLogger,
+        test_logger: *DirectPrintLogger,
+        noop: void,
+
+        const Self = @This();
+
+        /// Can be used in tests to minimize the amount of logging during tests.
+        pub const TEST_DEFAULT_LEVEL: Level = .warn;
+
+        pub fn unscoped(self: Self) Logger {
+            return switch (self) {
+                .standard => |logger| .{ .standard = logger },
+                .test_logger => |logger| .{ .test_logger = logger },
+                .noop => .noop,
+            };
         }
-    }
 
-    pub fn infof(self: Self, comptime fmt: []const u8, args: anytype) void {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                logger.infof(fmt, args);
-            },
-            .noop => {},
+        pub fn withScope(self: Self, comptime new_scope: []const u8) ScopedLogger(new_scope) {
+            return switch (self) {
+                .standard => |logger| .{ .standard = logger },
+                .test_logger => |logger| .{ .test_logger = logger },
+                .noop => .noop,
+            };
         }
-    }
 
-    pub fn debugf(self: Self, comptime fmt: []const u8, args: anytype) void {
-        switch (self) {
-            inline .standard, .test_logger => |logger| {
-                logger.debugf(fmt, args);
-            },
-            .noop => {},
+        pub fn deinit(self: *const Self) void {
+            switch (self.*) {
+                .standard => |*logger| {
+                    logger.*.deinit();
+                },
+                .test_logger => |*logger| {
+                    logger.*.deinit();
+                },
+                .noop => {},
+            }
```

Comment:
these types are already defined as pointers so we dont need to capture them as a pointer again (a stack copy is fine)

```suggestion
        pub fn deinit(self: *const Self) void {
            switch (self.*) {
                .standard => |logger| logger.deinit(),
                .test_logger => |logger| logger.deinit(),
                .noop => {},
            }
```

Reply:
[aee8d3c](https://github.com/Syndica/sig/pull/277/commits/aee8d3c5a8358064bac60cf1b41456e45d835bd2)